kafka-rest/src/main/java/io/confluent/kafkarest/KafkaRestConfig.java

```diff
@@ -365,6 +371,43 @@
       "Whether to enable REST Proxy V3 API. Default is true.";
   private static final boolean API_V3_ENABLE_DEFAULT = true;
 
+  public static final String RATE_LIMIT_ENABLE_CONFIG = "rate.limit.enable";
+  private static final String RATE_LIMIT_ENABLE_DOC =
+      "Whether to enable request rate-limiting. Default is false.";
+  private static final boolean RATE_LIMIT_ENABLE_DEFAULT = false;
+
+  public static final String RATE_LIMIT_BACKEND_CONFIG = "rate.limit.backend";
+  private static final String RATE_LIMIT_BACKEND_DOC =
+      "The rate-limiting backend to use. The options are 'guava' and 'resilience4j'. Default is "
+          + "'guava'.";
+  private static final String RATE_LIMIT_BACKEND_DEFAULT = "guava";
+
+  public static final String RATE_LIMIT_PERMITS_PER_SEC_CONFIG = "rate.limit.permits.per.sec";
+  private static final String RATE_LIMIT_PERMITS_PER_SEC_DOC =
```

Comment:
I like that it's possible to give different endpoints a different rate limit, but I think the hard work here is on the user to do the maths - our logic should be able to do it.

What we want is to allow the user to configure that endpointA can do N requests per second (which is the rate limit measurement we use elsewhere, rather than costs and permits) and endpointB can do N+5 requests per second, as that's the familiar way of thinking about rate limits.

Having both costs per endpoint and a permit to aim for makes the user do the calculation.

I wonder if we can have a default rate limit per second (as there is already) and then costs is renamed to eg endpoints.rate.limits, and then from a user point of view each endpoint just has a rate limit in the map specified.

Then permits could go.  We'd have to work out how to translate the users requests per second for an endpoint in the rateLimit(costs) method... but actually, isn't costs just the rate limit for the endpoint anyway (although I suspect I'm missing something in thinking through that)?

Reply:
I actually think it would be harder to explain it to users. When you say "endpointA can do N requests per second" what you really mean is "in the absence of any other requests, endpointA can do N requests per second, but if there are other concurrent requests, then go figure". It's very hard to encode what the total-permits is in this case. The cost/permit way is, in my opinion, the most natural way of thinking of rate-limiting/QoS. The math is easy to reason about, and not a huge problem.